### ISSUE TITLE
Unsafe Query Generation handling causes strong parameters fail on empty arrays

### ISSUE THREAD

**Issue by odedniv** [rails/rails#20567]
Per http://guides.rubyonrails.org/security.html#unsafe-query-generation, Rails now converts empty arrays received as ActionController parameters to nil. This causes strong parameters to fail when it previously didn't.

This doesn't fail:
```ruby
ActionController::Parameters.new(user: { parents_attributes: [] }).
  require(:user).permit(parents_attributes: [:name])
```

While this produces `ActionController::UnpermittedParameters: found unpermitted parameter: parents_attributes`:
```ruby
ActionController::Parameters.new(user: { parents_attributes: nil }).
  require(:user).permit(parents_attributes: [:name])
```

I'm pretty sure it's not the client's responsibility, but I'm not sure whether its strong parameters' responsibility or the security fix's. Seems to me that the security fix screws up the parameters so it should be changed, but on the other hand it can cause a security issue.

**Comment by odedniv** [comment]
This is still reproducible with the code in the original post on 4.2.5.

**Comment by rafaelfranca** [comment]
Is it still reproducible with Rails 5.0.0.beta1?

**Comment by odedniv** [comment]
I don't know, not using Rails 5, but it's just one line in the console to check.

**Comment by rafaelfranca** [comment]
Please, be my guest to try it and report back.

**Comment by prathamesh-sonpatki** [comment]
This works on Rails 5

```
$ rails c
Loading development environment (Rails 5.0.0.beta1)
>> Rails.version
=> "5.0.0.beta1"
>> ActionController::Parameters.new(user: { parents_attributes: nil }).
?>     require(:user).permit(parents_attributes: [:name])
=> {}
>> 
```

**Comment by rafaelfranca** [comment]
Thanks for confirming. I'm closing this issue because it is already fixed in Rails 5. We can't fix in Rails 4.2 because it would need to change how the security protection of Rails 4 and it would be dangerous to do in a stable release.
